### ISSUE TITLE
ASYNC240 false positive for `Path.glob()` iterator on offloaded thread

### ISSUE THREAD

**Issue by tjkuson** [astral-sh/ruff#28040]
### Summary

[blocking-path-method-in-async-function (ASYNC240)](https://docs.astral.sh/ruff/rules/blocking-path-method-in-async-function/#blocking-path-method-in-async-function-async240) reports on `pathlib.Path.glob` iterators even when passed to a worker thread for consumption.

```python
import asyncio
import pathlib

async def fn(directory: pathlib.Path) -> list[pathlib.Path]:
    return await asyncio.to_thread(list, directory.glob("*.json"))
```

Ruff reports the diagnostic on the usage of `directory.glob`. The expected behaviour is that no diagnostic is reported as the filesystem traversal occurs when the iterator is consumed. Here, `list` consumers the iterator in a worker thread created by `asyncio.to_thread` and thus does not block the event loop. 

### Version

ruff 0.16.4

**Comment by tjkuson** [comment]
What ought to be flagged is the _consumption_ of the `pathlib.Path.glob` object rather than its creation. Doing that correctly likely requires type analysis. For what it's worth, flake8-async doesn't report on `pathlib.Path.glob` objects at all or anything else that blocks only when iterated upon.

**Comment by ntBre** [comment]
Thanks, this seems like a bug, but I'm not sure how easily Ruff can understand this. We could add an exception for exactly this pattern (`.glob` directly in `asyncio.to_thread`), but that's pretty specific and leaves many variations out. I don't think we could trace the `glob` object to the actual iteration site in general, which would probably lead to us just not emitting diagnostics on `Path.glob`.

I don't think we can take inspiration from flake8-async here, it sounds like they don't implement any `pathlib` checks: https://github.com/python-trio/flake8-async/issues/396.

I think it might make sense to split `glob` detection into a separate rule if it's more pedantic than the other patterns the rule catches. This rule is currently in the `suspicious` category.

**Comment by tjkuson** [comment]
> I don't think we can take inspiration from flake8-async here, it sounds like they don't implement any `pathlib` checks: [python-trio/flake8-async#396](https://github.com/python-trio/flake8-async/issues/396).

I did see that, but for what it's worth, the draft PR https://github.com/python-trio/flake8-async/pull/467 doesn't report on `pathlib.Path.glob` either, nor does the upstream rule report any `os.path` object that block only upon iteration, so Ruff is still a bit unique here. I appreciate it's not clear-cut, though!

**Comment by ntBre** [comment]
Ah makes sense, thanks! I hadn't seen that PR. Seems like more evidence for splitting `glob` into its own rule.
